### ISSUE TITLE
file_limit management

### ISSUE THREAD

**Issue by 0x1793d1** [kardeiz/awmp#1]
Hi,

Thanks for your work on awmp.
However, I have a remark regarding the file limit management: Why discarding the additional bytes? Is there a way to return an error instead?

Thanks.

**Comment by kardeiz** [comment]
Hi, and thanks for the issue!

I might be misunderstanding your question, but I think discarding the additional bytes is the point of the `file_limit` configuration: if you have limited filesystem storage, for example, you don't want users to be able to upload multi-GB files. I've checked and it doesn't seem like discarding these bytes messes up the multipart `Stream` (that is, you can still access "parts" that come after the discarded part).

Please let me know if I misunderstand, or if you think something else should be done with the additional bytes.

Regarding the other part of your question (returning an error for too large file): sure, I think it could be helpful to return an error for those cases. I've put together a PR here: #2.

Basically, `FileParts(Vec<(String, File)>)` becomes `FileParts(Vec<(String, Result<File, Error>)>)`, where `Error` is `Error::FileTooLarge { limit: usize, file_name: Option<String> }`.

The convenience function `remove` (which I have soft renamed to `take`) still returns a `Vec<File>`, and a new convenience function `first` returns `Option<&File>`, since I think ignoring the error is the more common scenario. However, you can use `into_inner` to get the `Vec<(String, Result<File, Error>)>`. (This changes the signature for `into_inner`, which is why I've bumped the version to `0.4.0`.)

Please let me know here or on the PR if you have any questions or suggestions or would like to propose a different way to handle this.

Thanks!

**Comment by 0x1793d1** [comment]
Hi,

Thanks for your reply!
You understood my question correctly.

I want to return a different HttpResponse when a FileTooLarge error occurs.
I will try to do it with v0.4.0.

**Comment by kardeiz** [comment]
Published [`0.4.0`](https://crates.io/crates/awmp). Let me know if you have any questions or run into anything else!

**Comment by 0x1793d1** [comment]
Great!
Your help would be appreciated because I can't do what I want.

I'm trying to do the following:
- I get my uploaded file back
- If it exceeds the size limit, I want to return a `Result<HttpResponse>`
- Otherwise I want to process the file and return another `Result<HttpResponse>`

e.g. :
```rust
if let Ok(file) = ? {
    let filename = ?;
    ...
    Ok(HttpResponse::Found().header(LOCATION, "/").finish())
} else {
    Err(ErrorBadRequest("Uploaded file is too big"))
}
```

I tried with `parts.files.take("file").first()`, but It's impossible to take the ownership of the `file`.

BTW, I don't know if Bad Request is a good response for that kind of issue.

Thanks in advance!

**Comment by kardeiz** [comment]
Here is an example:

```rust
use actix_web::FromRequest;

pub fn upload(parts: awmp::Parts) -> Result<actix_web::HttpResponse, actix_web::Error> {
    let files = parts.files.into_inner();

    let file_or_err = files.into_iter()
        .filter(|(k, _)| k == "file")
        .map(|(_, v)| v)
        .next()
        .ok_or_else(|| actix_web::error::ErrorBadRequest("Must upload file"))?;

    match file_or_err {
        Ok(file) => {
            let filename = file.sanitized_file_name();
            // Do something with `file`
            Ok(actix_web::HttpResponse::Ok().body("File was processed"))
        },
        // By default this will return something like 
        // `File is too large (limit: 10 bytes): test.txt`
        Err(e) => Err(actix_web::error::ErrorBadRequest(e))
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    actix_web::HttpServer::new(move || {
        actix_web::App::new()
            .data(awmp::Parts::configure(|cfg| cfg.with_file_limit(10)))
            .route("/", actix_web::web::post().to(upload))
    })
    .bind("0.0.0.0:3000")?
    .run()?;

    Ok(())
}
```

If this is for a normal web interface, you would probably want to use a redirect + flash message rather than just return an `ErrorBadRequest`, but that is outside the scope of this library. For flash messages in `actix-web`, check out [`actix-web-flash`](https://github.com/hatzel/actix-web-flash), but make sure you use the git repo with `master` branch (which has a PR from me to update to actix-web 1.0), rather than the published version (which is for actix-web 0.7).

**Comment by 0x1793d1** [comment]
It worked, thanks a lot!!!
I will try actix-web-flash. Thank you for your help, and thanks again for your work!
